design/external-references.md

```diff
@@ -0,0 +1,344 @@
+    Type: AWS::IAM::Group
+```
+
+and:
+
+```yaml
+Resources:
+  User00B015A1:
+    Type: AWS::IAM::User
+    Properties:
+      Groups:
+      - Fn::ImportValue: ExportsOutputRefGroupC77FDACD8CF7DD5B
+```
+
+You can see that the reference automatically resulted in an `Output+Export` and
+an `Fn::ImportValue` to be used in order to share this information across the
+stacks within the same application.
+
+This mechanism practically replaces our "import/export" pattern which looks like this (same use case and same synthesized output):
+
+```ts
+const user = new iam.User(stackWithUser, 'User');
+const group = new iam.Group(stackWithGroup, 'Group');
+
+// creates an `Output` in "stackWithUser" with a unique export name
+// returns an `Fn::ImportValue` that can be used to import this user
+// in any stack in the same account/region
+const importValues = user.export();
+
+// instantiates an `IUser` whose attributes are `Fn::ImportValue` tokens
+// in "stackWithGroup", and it can now be used "naturally" in that stack
+const importedUser = iam.User.import(stackWithGroup, 'ImportedUser', importValues);
+group.addUser(importedUser);
+```
+
+Obvsiouly the world is a better place now. Thanks [@rix0rrr](https://github.com/users/rix0rrr).
+
+## Problem Statement
+
+So, "what seems to be the problem?", you ask. Well, we want more! The current model only supports references that occur between stacks within the same CDK app. What about the use case where I want to reference a resource that was created in a totally different app?
+
+A common example is for a common infrastructure team to maintain a CDK app which defines networking resources such as VPCs, security groups, LBs, etc and application teams need to reference these resources in their own CDK apps.
+
+__So the problem we are trying to solve is how to enable users to reference resources (and potentially higher level constructs) across CDK apps.__
+
+## Scope
+
+As an initial step, we'll address the use case of referencing resources across apps but within __the same AWS environment__ (account + region). This limitation is based on the limitation AWS CloudFormation has for exports/imports.
+
+However, there are definitely use cases for being able to reference resources across environments, and we are also aware that some customers are reluctant to use CloudFormation import/exports due to the strong-coupling this mechanism creates between the stacks (for example, if an Output of one stack is being referenced by another stack, the original stack cannot be deleted).
+
+Another related use case is referencing resources by runtime code. An application's runtime code (AWS Lambda functions, CodeBuild projects, ECS containers, etc) needs to be able to interact with resources defined in CDK apps (e.g. read/write from a Amazon DynamoDB table, send messages to an SQS Queue, etc). This is a very similar use case to the cross-app references because basically the same information needs to be transferred between the apps.
```

Comment:
I'm not sure I agree with this. For example, a Load Balancer has like 5 attributes that need to be transported for infra construction (among which a VPC and a list of security groups), but ultimately the runtime code is only interested in the name. I don't think the needs are similar at all (I will agree insofar as that "they're both ways of transporting information across process boundaries", but that's not saying a lot, that happens all the time)

Reply:
I was thinking in line with @sam-goodwin’s model where you can basically access CDK constructs at runtime. Granted, in most cases you don’t need all runtime attributes, but if you have an IVpcNetwork object in your hands, then technically you are able to reference any of those inner resources at runtime. Maybe that’s not always needed (or even never needed), but if we can easily and transparently make these attributes available at runtime, then maybe that’s not a problem.